Source/Fuse.Controls.Native/iOS/InputDispatch.uno

```diff
@@ -8,131 +8,197 @@ namespace Fuse.Controls.Native.iOS
+			Moved = 1,
+			Released = 2,
+			Cancelled = 3,
 		}
 
-		public static void HandleEvent(ObjC.Object viewHandle, Visual origin, ObjC.Object uiEvent)
+		static readonly Action<Visual,PointerEventData>[] _handlers = new Action<Visual,PointerEventData>[] {
+			RaisePressed,
+			RaiseMoved,
+			RaiseReleased,
+			RaiseCancelled,
+		};
+
+		static void RaiseEvent(EventType type, Visual origin, ObjC.Object touches)
 		{
-			if (uiEvent.IsUIEvent())
-				HandleEvent(viewHandle, origin, new UIEvent(uiEvent));
+			var rootInfo = GetRootInfo(origin);
+			var count = NSArrayCount(touches);
+			var handler = _handlers[(int)type];
+			var deactivate = ((int)type) > ((int)EventType.Moved);
+			for (var i = 0; i < count; i++)
+			{
+				var uiTouch = NSArrayObjectAtIndex(touches, i);
+				handler(rootInfo.RootVisual, NewPointerEventData(uiTouch, rootInfo.RootView));
+				if (deactivate)
+					DeactivateTouch(uiTouch);
+			}
 		}
 
-		public static void HandleEvent(ObjC.Object viewHandle, Visual origin, UIEvent uiEvent)
+		class RootInfo
 		{
-			var rootVisual = FindRoot(origin);
-			var touches = uiEvent.GetTouchesForView(viewHandle);
-			var touchEnded = false;
+			public readonly Visual RootVisual;
+			public readonly ObjC.Object RootView;
+
+			public RootInfo(Visual rootVisual, ObjC.Object rootView)
+			{
+				RootVisual = rootVisual;
+				RootView = rootView;
+			}
+		}
 
+		static RootInfo GetRootInfo(Visual origin)
+		{
+			var rootVisual = FindRoot(origin);
 			ObjC.Object rootView = null;
 			var rootViewport = rootVisual as NativeRootViewport;
 			if (rootViewport != null)
 				rootView = rootViewport.RootView.NativeHandle;
+			return new RootInfo(rootVisual, rootView);
+		}
 
-			for (var i = 0; i < touches.Length; i++)
-			{
-				var touch = touches[i];
+		const int PointerCount = 10;
 
-				ActivateTouch(touch);
-				var pointerIndex = GetPointerIndex(touch);
-				var data = MakePointerEventData(touch, rootView, pointerIndex);
+		static ObjC.Object[] _activePointers = new ObjC.Object[PointerCount];
 
-				if (touch.Phase == UITouch.TouchPhase.Began)
-				{
-					RaisePressed(rootVisual, origin, data);
-				}
-				else if (touch.Phase == UITouch.TouchPhase.Moved)
-				{
-					RaiseMoved(rootVisual, origin, data);
-				}
-				/*
-				else if (touch.Phase == UITouch.TouchPhase.Staionary) { }
-				*/
-				else if (touch.Phase == UITouch.TouchPhase.Ended)
-				{
-					RaiseReleased(rootVisual, origin, data);
-					touchEnded = true;
-				}
-				else if (touch.Phase == UITouch.TouchPhase.Cancelled)
-				{
-					RaiseCancelled(origin, data);
-					touchEnded = true;
-				}
+		static int GetPointIndex(ObjC.Object uiTouch)
+		{
+			var availableIndex = -1;
+			for (var i = PointerCount; i-- > 0;)
+			{
+				var pointer = _activePointers[i];
+				if (CompareUITouch(pointer, null))
+					availableIndex = i;
+				else if (CompareUITouch(pointer, uiTouch))
+					return i;
 			}
+			_activePointers[availableIndex] = uiTouch;
+			return availableIndex;
```

Comment:
Why does this scan backwards? If it scanned forwards, and saved the first available index instead of continiously overwriting, wouldn't the loop terminate earlier in the common-case of few pointers down? I also think the logic would be a bit more straight forward...